File: bench/src/cs/src/dijkstra.cs
using System;
using System.Diagnostics;

static class Bench
{
    const ulong N = 800;
    const ulong NN = N * N;
    const ulong INF = 0x0FFFFFFFFFFFFFFF;

    static ulong gSeed = 12345;

    static ulong Rnd()
    {
        gSeed = (gSeed * 16807) % 2147483647;
        return gSeed;
    }

    static ulong[] gHeapD;
    static ulong[] gHeapN;
    static ulong gHeapSize = 0;

    static ulong gPopD = 0;
    static ulong gPopN = 0;

    static void Push(ulong d, ulong node)
    {
        ulong i = gHeapSize;
        gHeapSize += 1;
        gHeapD[(int) i] = d;
        gHeapN[(int) i] = node;

        while (i > 0)
        {
            ulong p = (i - 1) >> 1;
            if (gHeapD[(int) p] <= gHeapD[(int) i])
                break;
            ulong td = gHeapD[(int) p];
            gHeapD[(int) p] = gHeapD[(int) i];
            gHeapD[(int) i] = td;
            ulong tn = gHeapN[(int) p];
            gHeapN[(int) p] = gHeapN[(int) i];
            gHeapN[(int) i] = tn;
            i = p;
        }
    }

    static void Pop()
    {
        gPopD = gHeapD[0];
        gPopN = gHeapN[0];
        gHeapSize -= 1;
        gHeapD[0] = gHeapD[(int) gHeapSize];
        gHeapN[0] = gHeapN[(int) gHeapSize];

        ulong i = 0;
        while (true)
        {
            ulong l = 2 * i + 1;
            if (l >= gHeapSize)
                break;
            ulong r = l + 1;
            ulong m = l;
            if (r < gHeapSize && gHeapD[(int) r] < gHeapD[(int) l])
                m = r;
            if (gHeapD[(int) i] <= gHeapD[(int) m])
                break;
            ulong td = gHeapD[(int) m];
            gHeapD[(int) m] = gHeapD[(int) i];
            gHeapD[(int) i] = td;
            ulong tn = gHeapN[(int) m];
            gHeapN[(int) m] = gHeapN[(int) i];
            gHeapN[(int) i] = tn;
            i = m;
        }
    }

    static int Main()
    {
        // ---- data generation (not timed) ----
        var weight = new ulong[NN];
        for (int i = 0; i < (int) NN; i++)
            weight[i] = 1 + (Rnd() % 9);

        // ---- timed work ----
        long start_t = Stopwatch.GetTimestamp();

        var dist = new ulong[NN];
        for (int i = 0; i < (int) NN; i++)
            dist[i] = INF;

        gHeapD = new ulong[NN * 4];
        gHeapN = new ulong[NN * 4];
        gHeapSize = 0;

        dist[0] = 0;
        Push(0, 0);

        ulong pops = 0;
        const ulong Target = NN - 1;

        while (gHeapSize > 0)
        {
            Pop();
            ulong d = gPopD;
            ulong u = gPopN;
            pops += 1;
            if (d > dist[(int) u])
                continue;
            if (u == Target)
                break;

            ulong x = u % N;
            ulong y = u / N;

            if (x > 0)
            {
                ulong v = u - 1;
                ulong nd = d + weight[(int) v];
                if (nd < dist[(int) v])
                {
                    dist[(int) v] = nd;
                    Push(nd, v);
                }
            }
            if (x < N - 1)
            {
                ulong v = u + 1;
                ulong nd = d + weight[(int) v];
                if (nd < dist[(int) v])
                {
                    dist[(int) v] = nd;
                    Push(nd, v);
                }
            }
            if (y > 0)
            {
                ulong v = u - N;
                ulong nd = d + weight[(int) v];
                if (nd < dist[(int) v])
                {
                    dist[(int) v] = nd;
                    Push(nd, v);
                }
            }
            if (y < N - 1)
            {
                ulong v = u + N;
                ulong nd = d + weight[(int) v];
                if (nd < dist[(int) v])
                {
                    dist[(int) v] = nd;
                    Push(nd, v);
                }
            }
        }

        ulong check = dist[(int) Target] * 1000 + (pops % 1000);

        double ms = (Stopwatch.GetTimestamp() - start_t) * 1000.0 / Stopwatch.Frequency;
        Console.WriteLine($"CHECK={check} MS={ms:F6}");
        return 0;
    }
}

File: bench/src/rust/dijkstra.rs
use std::time::Instant;

const N: u64 = 800;
const NN: u64 = N * N;
const INF: u64 = 0x0FFFFFFFFFFFFFFF;

static mut G_SEED: u64 = 12345;

fn rnd() -> u64 {
    unsafe {
        G_SEED = (G_SEED * 16807) % 2147483647;
        G_SEED
    }
}

struct Heap {
    d: Vec<u64>,
    n: Vec<u64>,
    size: u64,
}

impl Heap {
    fn push(&mut self, d: u64, node: u64) {
        let mut i = self.size;
        self.size += 1;
        self.d[i as usize] = d;
        self.n[i as usize] = node;

        while i > 0 {
            let p = (i - 1) >> 1;
            if self.d[p as usize] <= self.d[i as usize] {
                break;
            }
            self.d.swap(p as usize, i as usize);
            self.n.swap(p as usize, i as usize);
            i = p;
        }
    }

    fn pop(&mut self) -> (u64, u64) {
        let pop_d = self.d[0];
        let pop_n = self.n[0];
        self.size -= 1;
        self.d[0] = self.d[self.size as usize];
        self.n[0] = self.n[self.size as usize];

        let mut i = 0u64;
        loop {
            let l = 2 * i + 1;
            if l >= self.size {
                break;
            }
            let r = l + 1;
            let mut m = l;
            if r < self.size && self.d[r as usize] < self.d[l as usize] {
                m = r;
            }
            if self.d[i as usize] <= self.d[m as usize] {
                break;
            }
            self.d.swap(m as usize, i as usize);
            self.n.swap(m as usize, i as usize);
            i = m;
        }

        (pop_d, pop_n)
    }
}

fn main() {
    // ---- data generation (not timed) ----
    let mut weight = vec![0u64; NN as usize];
    for i in 0..NN as usize {
        weight[i] = 1 + (rnd() % 9);
    }

    // ---- timed work ----
    let start_t = Instant::now();

    let mut dist = vec![INF; NN as usize];

    let mut heap = Heap {
        d: vec![0; (NN * 4) as usize],
        n: vec![0; (NN * 4) as usize],
        size: 0,
    };

    dist[0] = 0;
    heap.push(0, 0);

    let mut pops: u64 = 0;
    let target = NN - 1;

    while heap.size > 0 {
        let (d, u) = heap.pop();
        pops += 1;
        if d > dist[u as usize] {
            continue;
        }
        if u == target {
            break;
        }

        let x = u % N;
        let y = u / N;

        if x > 0 {
            let v = u - 1;
            let nd = d + weight[v as usize];
            if nd < dist[v as usize] {
                dist[v as usize] = nd;
                heap.push(nd, v);
            }
        }
        if x < N - 1 {
            let v = u + 1;
            let nd = d + weight[v as usize];
            if nd < dist[v as usize] {
                dist[v as usize] = nd;
                heap.push(nd, v);
            }
        }
        if y > 0 {
            let v = u - N;
            let nd = d + weight[v as usize];
            if nd < dist[v as usize] {
                dist[v as usize] = nd;
                heap.push(nd, v);
            }
        }
        if y < N - 1 {
            let v = u + N;
            let nd = d + weight[v as usize];
            if nd < dist[v as usize] {
                dist[v as usize] = nd;
                heap.push(nd, v);
            }
        }
    }

    let check = dist[target as usize] * 1000 + (pops % 1000);

    let ms = start_t.elapsed().as_secs_f64() * 1000.0;
    println!("CHECK={} MS={:.6}", check, ms);
}
